### DISCUSSION TITLE
Preview mode: allow more space in the preview data being passed on

### DISCUSSION THREAD

**Discussion by fernandoabolafio** [vercel/next.js#32207] [Ideas]
# Bug report

This bug refers to the [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode).

## Describe the bug

When using the method `setPreviewData` with `preview` mode. There is a hard limit of 2kb for the data. When I look into [the code](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L407)  states the following:

```
  // limit preview mode cookie to 2KB since we shouldn't store too much
  // data here and browsers drop cookies over 4KB
```
Two things here stand as bugs for me:
1. The message here is misleading since the 2kb refers to the data passed as an argument to `setPreviewData` plus the data added by next. So, in reality, the data that can be passed by the user is way less than 2kb. Something around 1.2kb.
2. There is no need for a hard limit of 2kb for the total data if the idea is to have it below 4kb. 

## To Reproduce

Create an API function like this for example:
```ts
import type { NextApiRequest, NextApiResponse } from 'next';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  // Check the secret and next parameters
  res.setPreviewData({
    token:
      'eyJraWQiOiJmYnFoRGRlVk5DXC9SZnI2T0YyVTUrZWpmUEdKRDYzcW1Zd3YxZUZMcWs1TT0iLCJhbGciOiJSUzI1NiJ9.eyJhdF9oYXNoIjoicHhNeXFPWU5BUURKVUlXSnczeFhTZyIsInN1YiI6IjJlYjI0MTkzLWQ0OWYtNGI2Yi1iYWFiLWQ0MmI4ODNjOWVhOCIsImF1ZCI6IjEyZmU5aDN0Nm85bjVldWpoNW8wZ2I0YmRrIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImV2ZW50X2lkIjoiNGE4OTMzMjAtNzI2Ni00ZjYxLWJiYWUtYmQ1MjEwYjNiODdhIiwidG9rZW5fdXNlIjoiaWQiLCJhdXRoX3RpbWUiOjE2MDYzODE4NTQsImlzcyI6Imh0dHBzOlwvXC9jb2duaXRvLWlkcC51cy1lYXN0LTIuYW1hem9uYXdzLmNvbVwvdXMtZWFzdC0yX2o3VFYwT1lmNCIsImNvZ25pdG86dXNlcm5hbWUiOiJyYXNtdXMiLCJleHAiOjE2MDYzODU0NTQsImlhdCI6MTYwNjM4MTg1NCwiZW1haWwiOiJyYXNtdXNAc2hhcGUuZGsifQ.HJYIseoHDE5BkpWLK9MoXvGwlecm_EOgd4lSIPXMJwWwyePl699w1GlXqwtDLynpOYPrqMFNmEHy1anOTWlXBkNJXbVmANXwVH5yED4zbJbxl5P3g-9Jn_a5c5dAXYEJSGhfCzaqJK-se8rqPC7l6IQGoe6IKhNkZYYFCm56i2u1iaYzxDDKuMdxXTuhbC4zpvvWSM7v7AtPAH4EVlF6XnolzWblKzLi4K-OwIIJnOPyAefWxBoeL3Zsv7qbG570xcy7GArt6gOi5E4naQeALL7gAMcyqzkf2IPunkYaRrMvczVjYo_UT1N5jSo3gy5nwQZ6i-4E9WzQY50iizo9CA',
  });
};
```
## Expected behavior

- If the data provided to `sePreviewData` plus the data added by next is less than 4kb, no error should be thrown and the program should proceed.
- Otherwise, if the sum of the data exceeds the 4kb then a better message could state something like this:
> Preview data is limited to  **X** currently, reduce how much data you are storing as preview data to continue

Where X is `4kb - dataAddedByNext`. This can be an estimate if the data added by the next changes in size. It can also be calculated on the fly if no significant performance harm is added.

**Comment by shadcn** [comment]
Bumping this. 

It seems like the 2048 limit was an arbitrary value picked in https://github.com/vercel/next.js/pull/10831#discussion_r387845699

While this does prevent users from storing large amounts of data in the cookie, the 2048 is not enough for storing a JWT (signed with RS256), which seems like a pretty common case for making authenticated requests when in preview mode.

Can we revisit this?

**Comment by octavioamu** [comment]
this limit kind of invalidate the use of setPreviewData(data,{}) where data being a blog post for example. Just a few lines of text will exceed that limit. 

https://nextjs.org/docs/pages/building-your-application/configuring/preview-mode#previewdata-size-limits 
By design seems the idea is to pass a cookie, like an access token instead of the content to be rendered.
